### ISSUE TITLE
Showing full property path when using nested validators

### ISSUE THREAD

**Issue by vipetrul** [FluentValidation/FluentValidation#1906]
### Is your feature request related to a problem? Please describe.

Would love validation messages to display property full path when using nested validators. Currently when validating child objects using `setValidator()`, errors raised in child objects never reference parent property path, which as I understand is by design. Unless of course all validations are specified at the root object and use syntax like `RuleFor(x => x.prop.childProp)`.

Passing parent "context" to child validators via `setValidator(new ChildValidator("parentPropPath"))` works but it creates a lot of boilerplate code.

Perhaps I'm missing something obvious here?
Please advise.

### Describe the solution you'd like

Rendered errors from nested validators should display full property path.

**Comment by JeremySkinner** [comment]
Hi, this is intentional - the full path is available in the ValidationFailure's `PropertyName` property, while only the short version is included inside the error message. If you need to display the full path, you would need to explicitly make use of `ValidationFailure.PropertyName`.

**Comment by vipetrul** [comment]
Can't believe I missed that, thank you very much!

**Comment by gitfool** [comment]
I'm new to this awesome library and also need this functionality as I'm validating deeply nested configuration, so failures should display the _unmodified_ fully qualified configuration name.

@JeremySkinner I looked into your suggestion above and the problem with this approach is that [ValidationFailure](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/Results/ValidationFailure.cs#L27) doesn't track the _unformatted_ error message with the placeholder values, so the error message cannot be reconstructed as desired.

I had to work around two issues to get the desired formatting, which made me wondering if I was missing something:
1) [`SplitPascalCase`](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/Internal/ExtensionsInternal.cs#L59) is not configurable and used in [DefaultValidatorExtensions.GetDisplayName](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/DefaultValidatorExtensions.cs#L1228-L1229) and [RuleBase.PropertyName](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/Internal/RuleBase.cs#L190-L196)

The problem here is that this forces you to call [RuleBase.SetDisplayName](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/Internal/RuleBase.cs#L70-L87) or override the [Global](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/ValidatorOptions.cs#L126) [DisplayNameResolver](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/ValidatorOptions.cs#L85-L88):
```c#
ValidatorOptions.Global.DisplayNameResolver = (_, member, _) => member.Name; // avoids "split pascal case"
```
Okay, so it is configurable in the above way, just not as a separate option on the default display name resolver; but I was still surprised that there was a manipulation of the [property] name by default.

2) The property chain is reset in child validators

This is apparently by design, but unless there are technical reasons, it seemed to me that it should be more configurable. As it is, I had to override the display name _per rule_:
```c#
public sealed class ConfigValidator : AbstractValidator<Config>
{
    public ConfigValidator()
    {
        RuleFor(config => config.Environment.K8s.Flux).NotEmpty().When(config => config.Environment.K8s.InstallFlux)
            .ChildRules(validator =>
            {
                validator.RuleFor(flux => flux!.GitHubOwner).NotEmpty().WithFullName();
                validator.RuleFor(flux => flux!.GitHubRepository).NotEmpty().WithFullName();
                validator.RuleFor(flux => flux!.GitHubToken).NotEmpty().WithFullName();
                validator.RuleFor(flux => flux!.SlackUrl).NotEmpty().WithFullName();
            }).When(config => config.Environment.K8s.InstallFlux);
    }
}

...

public static class ValidationExtensions
{
    public static IRuleBuilderOptions<T, TProperty> WithFullName<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule)
    {
        rule.Configure(config => config.SetDisplayName(context => context.PropertyName));
        return rule;
    }
}
```
This would be more _easily_ configurable, avoiding repeated calls to `WithFullName`, if [DefaultValidatorExtensions.ChildRules](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/DefaultValidatorExtensions.cs#L1206) took an optional parameter that controlled whether or not to reset the property chain. Failing that, a way to configure a "scope" that would be inherited by all child rules would also help.

Looking forward to your thoughts as I might be missing something less obvious to a newcomer. 🙏

**Comment by gitfool** [comment]
Actually, there's another problem with (1) above. You can't just return the member name because that strips the property chain. [RuleBase.PropertyName](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/Internal/RuleBase.cs#L194) has the advantage of access to `_propertyName` when assigning `_propertyDisplayName`, so I had to copy the private implementation for [DefaultPropertyNameResolver](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/ValidatorOptions.cs#L103-L110):
```c#
public static class ValidationHelpers
{
    public static string DefaultPropertyNameResolver(Type type, MemberInfo member, LambdaExpression? expression)
    {
        if (expression != null)
        {
            var chain = PropertyChain.FromExpression(expression);
            if (chain.Count > 0) return chain.ToString();
        }
        return member.Name;
    }
}
```
and used that as the global `DisplayNameResolver`:
```c#
ValidatorOptions.Global.DisplayNameResolver = ValidationHelpers.DefaultPropertyNameResolver; // property chain
```
TL;DR: it definitely feels harder than it should be to deal with fully qualified property chains. 😞

**Comment by JeremySkinner** [comment]
Hi @gitfool 

The display name is intentionally intended to be end-user readable which is why it doesn't include the full path, and goes through SplitPascalCase, and is why the path is stored separately and isn't included in the message. This covers a common set of scenarios and is the 'standard' workflow.

If you need to include the full path as the display name then either approach you've taken is valid/correct; either by using an extension method on the rule or a global display name resolver that makes use of `PropertyChain.FromExpression`. Personally I prefer using the rule-level extension method as I think it's better to make it clear/explicit that you're diverging from the standard approach (although the global option is perfectly valid too).

Both these extensibility points are here exactly for this purpose - to allow you to change the default behaviour of the library if you need to, and it seems you've been able to do exactly what you need, so I don't think this requires any changes to the library.

**Comment by gitfool** [comment]
@JeremySkinner thanks for your reply that confirms my approaches. I was hoping there would be other less verbose ways, which ultimately might just be through providing the unformatted message in the result. Would that be possible?

**Comment by JeremySkinner** [comment]
That's not something I'm keen on, sorry. You'll need to either use one of the extensibility mechanisms mentioned above.

**Comment by gitfool** [comment]
@JeremySkinner fair enough. Meanwhile, I've worked out how to do all the processing on the [ValidationFailure](https://github.com/FluentValidation/FluentValidation/blob/6805642f445fbd6048d47954349d59cf8ff07eb8/src/FluentValidation/Results/ValidationFailure.cs#L27) such that I don't care about the display name, so I can drop both the global `DisplayNameResolver` override and `WithFullName` extension above:
```c#
public static class ConfigExtensions
{
    public static ValidationResult Validate(this Config config)
    {
        var result = new ConfigValidator().Validate(config);
        foreach (var error in result.Errors)
        {
            // replace split pascal case display name with fully chained property name
            var displayName = error.FormattedMessagePlaceholderValues["PropertyName"];
            error.ErrorMessage = error.ErrorMessage.Replace($"'{displayName}'", error.PropertyName); // strip quotes too ;)
        }
        return result;
    }
}
```

**Comment by gitfool** [comment]
The more I play with this library the more I'm impressed! It's extensible enough that I worked out how to validate dictionaries with string keys, where the kvp key should be a name in the property chain and the kvp value should be lifted into the child rules:
```c#
public sealed class ConfigValidator : AbstractValidator<Config>
{
    public ConfigValidator()
    {
        RuleFor(config => config.Environment.Aws.Eks.Addons).ForEach(addons => addons
            .KeyNameIndexer()
            .ChildRules(validator =>
            {
                validator.RuleFor(addon => addon.Name).NotEmpty();
                validator.RuleFor(addon => addon.Version).NotEmpty();
            }));
    }
}
```
... where the `Environment.Aws.Eks.Addons` property is of type `public Dictionary<string, AwsEksAddonConfig>`, using the following extensions and helpers:
```c#
public static class ValidationExtensions
{
    public static IRuleBuilderInitialCollection<IEnumerable<KeyValuePair<string, TValue>>, KeyValuePair<string, TValue>> KeyNameIndexer<TValue>(this IRuleBuilderInitialCollection<IEnumerable<KeyValuePair<string, TValue>>, KeyValuePair<string, TValue>> rule)
    {
        rule.OverrideIndexer((_, _, element, _) =>
        {
            var asProperty = Regex.IsMatch(element.Key, @"^[A-Z][A-Za-z]*$");
            return asProperty ? $".{element.Key}" : $"['{element.Key}']";
        });
        return rule;
    }

    public static IRuleBuilderInitial<KeyValuePair<string, TValue>, TProperty> RuleFor<TValue, TProperty>(this InlineValidator<KeyValuePair<string, TValue>> validator, Expression<Func<TValue, TProperty>> expression) =>
        validator.RuleFor(ChainProperties<KeyValuePair<string, TValue>, TValue, TProperty>(entry => entry.Value, expression))
            .Configure(config => config.PropertyName = config.Member.Name);

    private static Expression<Func<A, C>> ChainProperties<A, B, C>(Expression<Func<A, B>> inner, Expression<Func<B, C>> outer) =>
        Expression.Lambda<Func<A, C>>(ExpressionHelpers.ReplaceExpressions(outer.Body, outer.Parameters[0], inner.Body), inner.Parameters[0]);
}

public static class ExpressionHelpers
{
    public static T ReplaceExpressions<T>(T expression, Expression from, Expression to) where T : Expression =>
        new ExpressionReplacer { From = from, To = to }.VisitAndConvert(expression, nameof(ReplaceExpressions));

    private sealed class ExpressionReplacer : ExpressionVisitor
    {
        public override Expression? Visit(Expression? node) => node != null && node == From ? To : base.Visit(node);

        public Expression From { get; init; } = null!;
        public Expression To { get; init; } = null!;
    }
}
```
